### ISSUE TITLE
ensemble.stdev feature request

### ISSUE THREAD

**Issue by Jashaviti** [pmlmodelling/nctoolkit#5]
Hi, 
currently the ds.ensemble feature is able to compute easily the mean, max and min. 
I would like to request that standard deviation be included to provide a form of variability. 

thanks.

**Comment by r4ecology** [comment]
I have just added `ensemble_stdev` and `ensemble_var` methods: [commit here](https://github.com/pmlmodelling/nctoolkit/commit/0aec365087d19f6f4a0f4eb077fd10103565b36b). I have added tests for them, so things should work fine.

Right now these are vanilla operations and accept no arguments, so they will only calculate the standard deviation and variance at each time step.

You just need to do one of the following:

    ds.ensemble_stdev()
    ds.ensemble_var()

I've chosen not to provide an option to use nco, because there appears to be no computational benefit, unlike the other ensemble operations. The `ignore_time` argument available for other ensemble methods is not available yet. That can be achieved as follows:

    ds.merge_time()
    ds.tmean()

I have also renamed the temporal method `tvariance` to `tvar` and deprecated `tvariance` for consistency.

I will push these updates to pypi and conda early next week.

**Comment by Jashaviti** [comment]
This is great. Thanks a whole bunch. Looking forward.

**Comment by r4ecology** [comment]
This has now been pushed to pypi and conda-forge.

**Comment by Jashaviti** [comment]
Awesome, thanks I will get to use it over the next few days. Thanks

**Comment by Jashaviti** [comment]
Bula Robert.

Thank you again for the updates in version 7. I have used them already on
multiple occasions, they are working very well and have helped me progress
quite a bit.

I have another question. This is something I was able to 'untidily' scratch
together without using nctoolkit and am wondering if there is a cleaner
more efficient/consistent way of doing this using nctoolkit.

Is there a way to count within each grid the consecutive and total number
of days (or weeks/months) spent above a threshold value. For example in my
case at the moment, I have average daily temperature values for the span of
several years at a 0.01 degree grid. or each grid I want  to
1. Identify the day (timestamp) where average daily temp was first above a
threshold temp (say 28 deg_celc).  Each time this is identified, count how
many consecutive days spent at or above the threshold. Each time the count
exceeds a certain number of days (say 14 days) and 'event' is recorded.
Using this, be able to plot a heatmap with 'hotter' zones referring to the
number of 'events'.
2.  count the overall number of days spent above the threshold temp.

Please let me know if there is a workaround using nctoolkit. Also, please
let me know if you prefer me asking this directly within stack overflow
instead.

Thank you and kind regards from a very rainy 'cold' Fiji.
Jasha

**Comment by Jashaviti** [comment]
Hi Robert,
Thank you for this. I think I understand the logic  of combining these
commands. I will see how I can integrate it into my workflow.
I will keep my eyes and ears open for any updates you release for
calculating consecutive days above a threshold.

Thanks again.
Jasha

**Comment by Jashaviti** [comment]
Bula again Robert,

sorry to again be emailing you for help with nctookit. Please do let me
know if you prefer to write directly in stacks.

This time my question is not based on identifying events above a threshold
as was our previous topic, nstead this time its to do with
merging/appending different grids.

I recently downloaded salinity data from opendap for Tuvalu, Wallis and
Futuna and Fiji -->
https://podaac.jpl.nasa.gov/dataset/OISSS_L4_multimission_7day_v1?ids=&values=&search=salinity&provider=PODAAC

I had to download the data in 2 sections, once for everything east on 180
degrees  ( 176 E to 180 E) and once for everything west of 180 (176 W to
180W). I figured I could convert long to 360 and merge it using nc.merge. This
is however proving to be harder than planned as eventually, it clicked to
me that  nc.merge, is just for merging two datasets along the same grid,
and not to append a second gird and variables to the first grid.  (nc.merge
will ask that the second dataset be re-gridded to the first). I've skimmed
through the user guide and found nc.append.  I tried  something along the
lines of
ds1 = nc.open_data(eastdata)
ds2 = nc.open_data(westdata)

ds1.append(ds2)
ds1.tmean()

ds1.to_nc('tmeanofappend.nc')

everything seems to work until I get to ds1.to_nc..... error is 'You cannot
save multiple files'

Do you have any suggestions on how I can go about merging the two different
grids in a way i can process and export the output? Is this doable with
nctoolkit?

Thank you again for all the help, If there is anything I can do in return,
please do let me know.

Kind regards,
Jasha

**Comment by r4ecology** [comment]
Merging grids is possible in nctoolkit, though right now it's poorly documented. You can possibly do this using the `collect` method. This will merge netCDF files into 1, so long as the merged grid is complete. For example, if you had 2 files with one being the northern hemisphere and the other the south, you could use `collect` to get the entire globe. The separate netCDF files also need to have grids of the same shape.

In your case it should work, so long as the coordinates do not overlap, which they don't seem to. Try the following:

    ds1 = nc.open_data(eastdata)
    ds2 = nc.open_data(westdata)

    ds1.append(ds2)
    ds1.tmean()

    ds1.collect()

    ds1.to_nc('tmeanofappend.nc')

**Comment by Jashaviti** [comment]
Dear Robert,
Bula from Fiji,

A few days ago I updated nctoolkit to 0.4.0 from 0.3.7. Most things work as
very well, as usual. But I seem to be having issues with ds.plot. It seems
that one of the modules? is missing.

I tried re-installing, creating new virtual environments with a direct new
install of nctoolkit, using different data files, code, etc. The error
message remains the same regardless of what I do. However, when I revert
back to v0.3.7, everything works fine.

Here an example of the code I run and the error.

eg, code

ds = nc.open_data('/home/jasha/Data/ghrsst_2002_2020')
ds.select(year = 2006)ds.tmean()ds.plot()

output

ds.plot()
Traceback (most recent call last):

  Input In [18] in <module>
    ds.plot()

  File ~/anaconda3/envs/ncenv/lib/python3.8/site-packages/nctoolkit/plot.py:2
in plot
    from ncplot import view

ImportError: cannot import name 'view' from 'ncplot'
(/home/jasha/anaconda3/envs/ncenv/lib/python3.8/site-packages/ncplot/__init__.py)


Is this a common problem with the latest version? or is it just me?

Any insights into how I can go about fixing it are appreciated.

Vinaka,

Jasha

**Comment by Jashaviti** [comment]
Sorry, I should also mention that I have updated/reinstalled/ ncplot
directly myself rather than depending on the auto install when creating the
environment/installing nctoolkit

**Comment by r4ecology** [comment]
Hi

There is a present issue with the development version not working on some systems [https://github.com/pmlmodelling/nctoolkit/issues/14](https://github.com/pmlmodelling/nctoolkit/issues/14) due to a recent update to a pandas dependency.

What version of ncplot have you installed? Conda is possibly going to an old version. Though v 0.0.3 should be the only version without view, so I'm surprised it's going that early.

I think that's a rare event, but I've just changed the conda feedstock of nctoolkit so it will install a compatible version of ncplot by default, and I've modified the requirements.txt file on GitHub [https://github.com/pmlmodelling/nctoolkit/commit/bb83a1843b996d3e4701568b9aa4d0ec012d2308](https://github.com/pmlmodelling/nctoolkit/commit/bb83a1843b996d3e4701568b9aa4d0ec012d2308) so that ncplot should have a later pinned version. Hopefully that will resolve the problem.
